core/src/test/java/io/grpc/internal/ServiceConfigErrorHandlingTest.java

```diff
@@ -0,0 +1,627 @@
+    checkState(channel == null);
+
+    channel =
+        new ManagedChannelImpl(
+            channelBuilder,
+            mockTransportFactory,
+            new FakeBackoffPolicyProvider(),
+            balancerRpcExecutorPool,
+            timer.getStopwatchSupplier(),
+            Arrays.asList(interceptors),
+            timer.getTimeProvider());
+
+    int numExpectedTasks = 0;
+
+    // Force-exit the initial idle-mode
+    channel.syncContext.execute(new Runnable() {
+      @Override
+      public void run() {
+        channel.exitIdleMode();
+      }
+    });
+    if (channelBuilder.idleTimeoutMillis != ManagedChannelImpl.IDLE_TIMEOUT_MILLIS_DISABLE) {
+      numExpectedTasks += 1;
+    }
+
+    if (getNameResolverRefresh() != null) {
+      numExpectedTasks += 1;
+    }
+
+    assertEquals(numExpectedTasks, timer.numPendingTasks());
+
+    ArgumentCaptor<Helper> helperCaptor = ArgumentCaptor.forClass(null);
+    verify(mockLoadBalancerProvider).newLoadBalancer(helperCaptor.capture());
+  }
+
+  @Before
+  public void setUp() throws Exception {
+    when(mockLoadBalancer.canHandleEmptyAddressListFromNameResolution()).thenCallRealMethod();
+    LoadBalancerRegistry.getDefaultRegistry().register(mockLoadBalancerProvider);
+    expectedUri = new URI(TARGET);
+    when(mockTransportFactory.getScheduledExecutorService())
+        .thenReturn(timer.getScheduledExecutorService());
+    when(executorPool.getObject()).thenReturn(executor.getScheduledExecutorService());
+
+    channelBuilder =
+        new ChannelBuilder()
+            .nameResolverFactory(new FakeNameResolverFactory.Builder(expectedUri).build())
+            .defaultLoadBalancingPolicy(MOCK_POLICY_NAME)
+            .userAgent(USER_AGENT)
+            .idleTimeout(
+                AbstractManagedChannelImplBuilder.IDLE_MODE_MAX_TIMEOUT_DAYS, TimeUnit.DAYS)
+            .offloadExecutor(blockingExecutor);
+    channelBuilder.executorPool = executorPool;
+    channelBuilder.binlog = null;
+    channelBuilder.channelz = channelz;
+  }
+
+  @After
+  public void allPendingTasksAreRun() throws Exception {
+    // The "never" verifications in the tests only hold up if all due tasks are done.
+    // As for timer, although there may be scheduled tasks in a future time, since we don't test
+    // any time-related behavior in this test suite, we only care the tasks that are due. This
+    // would ignore any time-sensitive tasks, e.g., back-off and the idle timer.
+    assertTrue(timer.getDueTasks() + " should be empty", timer.getDueTasks().isEmpty());
+    assertEquals(executor.getPendingTasks() + " should be empty", 0, executor.numPendingTasks());
+    if (channel != null) {
+      channel.shutdownNow();
+      channel = null;
+    }
+  }
+
+  @After
+  public void cleanUp() {
+    LoadBalancerRegistry.getDefaultRegistry().deregister(mockLoadBalancerProvider);
+  }
+
+  @Test
+  public void emptyAddressesWithValidConfig_lbNeedsAddress() throws Exception {
```

Comment:
This is actually `emptyAddressesWithValidConfig_lbNeedsAddress_firstResolution()`, and we also need a `emptyAddressesWithValidConfig_lbNeedsAddress_secondResolution()` where the config is ignored.

Please make sure to cover all the cases listed under "Service Config Handling" in the spreadsheet.

It would be better to standardize the naming of the test cases according to the columns from the spec spreadsheet, so that we can easily check the test methods against the spec. For example, this would be `emptyAddresses_validConfig_firstResolution_lbNeedsAddress()`.

Reply:
tests are added. PTALA